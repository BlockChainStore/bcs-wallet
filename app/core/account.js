// @flow
import { wallet } from 'neon-js'

import { getStorage, setStorage } from './storage'

function getNEP6AddressData (matchesEncryptedWIF: boolean, address: string) {
  if (matchesEncryptedWIF) {
    return { address }
  } else {
    return {}
  }
}

export async function upgradeNEP6AddAddresses (encryptedWIF: string, wif: string) {
  const data = getStorage('userWallet')
  const loggedIntoAccount = new wallet.Account(wif)

  if (data && data.accounts) {
    const accounts = data.accounts.map((account, idx) => ({
      ...account,
      ...getNEP6AddressData(account.key === encryptedWIF, loggedIntoAccount.address)
    }))

    await setStorage('userWallet', { ...data, accounts })
  }
}
