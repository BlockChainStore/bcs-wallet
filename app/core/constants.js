// @flow
export const NEON_WALLET_RELEASE_LINK =
  'https://github.com/CityOfZion/neon-wallet/releases'

export const ASSETS = {
  NEO: 'NEO',
  GAS: 'GAS'
}

export const EXPLORERS = {
  NEO_TRACKER: 'Neotracker',
  NEO_SCAN: 'Neoscan',
  ANT_CHAIN: 'Antchain',
  NEO_VERSE: 'NEOverse'
}

export const ROUTES = {
  HOME: '',
  DASHBOARD: '/dashboard',
  CREATE_WALLET: '/create',
  ENCRYPT_KEY: '/encrypt-key',
  LOGIN_PRIVATE_KEY: '/login-private-key',
  LOGIN_LOCAL_STORAGE: '/login-local-storage',
  LOGIN_LEDGER_NANO_S: '/login-ledger-nano-s',
  LOGIN_NEP2: '/login-encrypted',
  LOGIN_TOKEN_SALE: '/login-token-sale',
  TOKEN_SALE: '/token-sale',
  SETTINGS: '/settings',
  DISPLAY_WALLET_KEYS: '/display-wallet-keys'
}

export const NOTIFICATION_LEVELS = {
  ERROR: 'error',
  SUCCESS: 'success',
  INFO: 'info',
  WARNING: 'warning'
}

export const NOTIFICATION_POSITIONS = {
  TOP_CENTER: 'tc',
  TOP_RIGHT: 'tr',
  TOP_LEFT: 'tl',
  BOTTOM_CENTER: 'bc',
  BOTTOM_RIGHT: 'br',
  BOTTOM_LEFT: 'bl'
}

export const BIP44_PATH =
  '8000002C' + '80000378' + '80000000' + '00000000' + '00000000'

export const MODAL_TYPES = {
  SEND: 'SEND',
  RECEIVE: 'RECEIVE',
  CONFIRM: 'CONFIRM',
  TOKEN_INFO: 'TOKEN_INFO',
  TOKEN: 'TOKEN',
  ICO: 'ICO'
}

export const MAIN_NETWORK_ID = '1'
export const TEST_NETWORK_ID = '2'

// TestNet
export const TOKENS_TEST = {
  BCS : 'e81c837d37763afbf894d319c868a09f219f2be9'
}

// MainNet
export const TOKENS = {
  BCS : '546c5872a992b2754ef327154f4c119baabff65f'
}

export const DEFAULT_WALLET = {
  name: 'userWallet',
  version: '1.0',
  scrypt: {
    cost: 16384,
    blockSize: 8,
    parallel: 8,
    size: 64
  },
  accounts: [],
  extra: null
}

export const DEFAULT_CURRENCY_CODE = 'usd'

export const CURRENCIES = {
  aud: { symbol: '$' },
  brl: { symbol: 'R$' },
  cad: { symbol: '$' },
  chf: { symbol: 'Fr.' },
  clp: { symbol: '$' },
  cny: { symbol: '¥' },
  czk: { symbol: 'Kč' },
  dkk: { symbol: 'kr. ' },
  eur: { symbol: '€' },
  gbp: { symbol: '£' },
  hkd: { symbol: '$' },
  huf: { symbol: 'Ft ' },
  idr: { symbol: 'Rp ' },
  ils: { symbol: '₪' },
  inr: { symbol: '₹' },
  jpy: { symbol: '¥' },
  krw: { symbol: '₩' },
  mxn: { symbol: '$' },
  myr: { symbol: 'RM' },
  nok: { symbol: 'kr ' },
  nzd: { symbol: '$' },
  php: { symbol: '₱' },
  pkr: { symbol: '₨ ' },
  pln: { symbol: 'zł' },
  rub: { symbol: '₽' },
  sek: { symbol: 'kr ' },
  sgd: { symbol: 'S$' },
  thb: { symbol: '฿' },
  try: { symbol: '₺' },
  twd: { symbol: 'NT$' },
  usd: { symbol: '$' },
  zar: { symbol: 'R ' }
}
