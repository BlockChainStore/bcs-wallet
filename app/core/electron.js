// @flow
import { shell } from 'electron'

export const openExternal = (srcLink: string) => {
  shell.openExternal(srcLink)
}
