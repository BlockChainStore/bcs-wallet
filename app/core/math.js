// @flow
import BigNumber from 'bignumber.js'

// https://stackoverflow.com/questions/4912788/truncate-not-round-off-decimal-numbers-in-javascript
export const truncateNumber = (num: number, places: number): number =>
  Math.trunc(num * 10 ** places) / 10 ** places

// https://github.com/MikeMcl/bignumber.js/issues/11
export const toBigNumber = (value: number | string) =>
  new BigNumber(String(value))

export const toNumber = (value: string | number) =>
  toBigNumber(value).toNumber()

export const isZero = (amount: string | number) =>
  toBigNumber(amount).equals(0)

export const isNumber = (value: string | number): boolean => {
  try {
    toBigNumber(value)
    return true
  } catch (e) {
    return false
  }
}
