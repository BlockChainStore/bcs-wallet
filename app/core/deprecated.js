// @flow
import { get, omit, pick } from 'lodash'

import { ID as AUTH_ID } from '../actions/authActions'
import { ID as BALANCES_ID } from '../actions/balancesActions'
import { ID as NETWORK_ID } from '../actions/networkActions'
import { ASSETS } from '../core/constants'
import { findNetwork } from '../core/networks'

const PREFIX = 'api'

// TODO: Module functions that inspect state such as `doClaimNotify` should be refactored into
//       higher-order components that expose function props for performing these actions.  Data that
//       is normally provided by functions like `getNetwork` can instead be retrieved via the
//       `withData` higher-order component.

export const getNetworkId = (state: Object) => {
  return get(state, `${PREFIX}.${NETWORK_ID}.data`)
}

export const getNetwork = (state: Object) => {
  return getNetworkById(getNetworkId(state))
}

export const getNetworkById = (networkId: string) => {
  return findNetwork(networkId).network
}

export const getAddress = (state: Object) => {
  return get(state, `${PREFIX}.${AUTH_ID}.data.address`)
}

export const getWIF = (state: Object) => {
  return get(state, `${PREFIX}.${AUTH_ID}.data.wif`)
}

export const getSigningFunction = (state: Object) => {
  return get(state, `${PREFIX}.${AUTH_ID}.data.signingFunction`)
}

export const getPublicKey = (state: Object) => {
  return get(state, `${PREFIX}.${AUTH_ID}.data.publicKey`)
}

export const getIsHardwareLogin = (state: Object) => {
  return get(state, `${PREFIX}.${AUTH_ID}.data.isHardwareLogin`)
}

export const getNEO = (state: Object): string => {
  return getBalances(state)[ASSETS.NEO]
}

export const getGAS = (state: Object): string => {
  return getBalances(state)[ASSETS.GAS]
}

export const getTokenBalances = (state: Object): Object => {
  return omit(getBalances(state), ASSETS.NEO, ASSETS.GAS)
}

export const getAssetBalances = (state: Object): Object => {
  return pick(getBalances(state), ASSETS.NEO, ASSETS.GAS)
}

export const getBalances = (state: Object) => {
  return get(state, `${PREFIX}.${BALANCES_ID}.data`)
}
