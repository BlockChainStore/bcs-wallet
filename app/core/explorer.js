// @flow
import { openExternal } from './electron'
import { EXPLORERS } from '../core/constants'
import { isMainNetwork } from '../core/networks'

export const getExplorerBaseURL = (networkId: string, explorer: ExplorerType) => {
  let baseURL
  if (explorer === EXPLORERS.NEO_TRACKER) {
    if (isMainNetwork(networkId)) {
      baseURL = 'https://neotracker.io'
    } else {
      baseURL = 'https://testnet.neotracker.io'
    }
  } else if (explorer === EXPLORERS.NEO_SCAN) {
    if (isMainNetwork(networkId)) {
      baseURL = 'https://neoscan.io'
    } else {
      baseURL = 'https://neoscan-testnet.io'
    }
  } else if (explorer === EXPLORERS.NEO_VERSE) {
    if (isMainNetwork(networkId)) {
      baseURL = 'http://explorer.neoverse.io'
    } else {
      baseURL = 'http://testnet.neoverse.io'
    }
  } else {
    if (isMainNetwork(networkId)) {
      baseURL = 'http://antcha.in'
    } else {
      baseURL = 'http://testnet.antcha.in'
    }
  }
  return baseURL
}

export const getExplorerTxLink = (networkId: string, explorer: ExplorerType, txId: string) => {
  const baseURL = getExplorerBaseURL(networkId, explorer)

  if (explorer === EXPLORERS.NEO_TRACKER) {
    return `${baseURL}/tx/${txId}`
  } else if (explorer === EXPLORERS.NEO_SCAN) {
    return `${baseURL}/transaction/${txId}`
  } else if (explorer === EXPLORERS.NEO_VERSE) {
    return `${baseURL}/transactions/${txId}`
  } else {
    return `${baseURL}/tx/hash/${txId}`
  }
}

export const getExplorerAddressLink = (networkId: string, explorer: ExplorerType, address: string) => {
  const baseURL = getExplorerBaseURL(networkId, explorer)

  if (explorer === EXPLORERS.NEO_TRACKER) {
    return `${baseURL}/address/${address}`
  } else if (explorer === EXPLORERS.NEO_SCAN) {
    return `${baseURL}/address/${address}/1`
  } else if (explorer === EXPLORERS.NEO_VERSE) {
    return `${baseURL}/addresses/${address}`
  } else {
    return `${baseURL}/address/info/${address}`
  }
}

export const getExplorerAssetLink = (networkId: string, explorer: ExplorerType, assetId: string) => {
  const baseURL = getExplorerBaseURL(networkId, explorer)

  if (explorer === EXPLORERS.NEO_TRACKER) {
    return `${baseURL}/asset/${assetId}`
  } else if (explorer === EXPLORERS.NEO_SCAN) {
    return `${baseURL}/asset/${assetId}`
  } else if (explorer === EXPLORERS.NEO_VERSE) {
    return `${baseURL}/assets/${assetId}`
  } else {
    return `${baseURL}/asset/hash/${assetId}`
  }
}

export const openExplorerTx = (networkId: string, explorer: ExplorerType, txId: string) =>
  openExternal(getExplorerTxLink(networkId, explorer, txId))

export const openExplorerAddress = (networkId: string, explorer: ExplorerType, address: string) =>
  openExternal(getExplorerAddressLink(networkId, explorer, address))

export const openExplorerAsset = (networkId: string, explorer: ExplorerType, assetId: string) =>
  openExternal(getExplorerAssetLink(networkId, explorer, assetId))
