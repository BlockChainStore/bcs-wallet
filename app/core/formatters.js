// @flow
import { ASSETS } from './constants'
import { toBigNumber } from './math'

export const COIN_DECIMAL_LENGTH = 8
export const SHORT_DISPLAY_DECIMAL_LENGTH = 4

type ValueType = string | number

export const formatGAS = (
  value: ValueType,
  shortDisplay: boolean = false
): string => {
  const decimals = shortDisplay
    ? SHORT_DISPLAY_DECIMAL_LENGTH
    : COIN_DECIMAL_LENGTH
  return toBigNumber(value).toFormat(decimals)
}

export const formatThousands = (value: ValueType): string =>
  toBigNumber(value).toFormat(0)

export const formatNEO = (value: ValueType): string =>
  toBigNumber(value).toFormat(0)

export const formatBalance = (
  symbol: SymbolType,
  balance: ValueType,
  shortDisplay: boolean = false
): string => {
  if (symbol === ASSETS.NEO) {
    return formatNEO(balance)
  }
  return formatGAS(balance, shortDisplay)
}

export const toFixedDecimals = (value: ValueType, decimals: number): string =>
  toBigNumber(value).toFixed(COIN_DECIMAL_LENGTH)

export const formatFiat = (value: ValueType): string =>
  toBigNumber(value).toFormat(2)
