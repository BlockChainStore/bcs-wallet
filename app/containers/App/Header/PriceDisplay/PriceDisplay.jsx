// @flow
import React from 'react'

import { formatFiat } from '../../../../core/formatters'
import { CURRENCIES } from '../../../../core/constants'

import headerStyles from '../Header.scss'

type Props = {
  neoPrice: number,
  gasPrice: number,
  currencyCode: string
}

const PriceDisplay = ({ neoPrice, gasPrice, currencyCode }: Props) => {
  let neoDisplayPrice = '-'
  let gasDisplayPrice = '-'
  const currencySymbol = CURRENCIES[currencyCode].symbol

  if (neoPrice) {
    neoDisplayPrice = formatFiat(neoPrice)
  }

  if (gasPrice) {
    gasDisplayPrice = formatFiat(gasPrice)
  }

  return [
    <div key='neoPriceDisplay' id='neoPrice' className={headerStyles.navBarItem}>
      <span className={headerStyles.navBarItemLabel}>NEO</span>
      <span className='price'>{currencySymbol}{neoDisplayPrice}</span>
    </div>,
    <div key='gasPriceDisplay' id='gasPrice' className={headerStyles.navBarItem}>
      <span className={headerStyles.navBarItemLabel}>GAS</span>
      <span className='price'>{currencySymbol}{gasDisplayPrice}</span>
    </div>
  ]
}

export default PriceDisplay
