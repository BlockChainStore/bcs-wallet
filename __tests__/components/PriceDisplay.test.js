import React from 'react'
import { mount } from 'enzyme'

import PriceDisplay from '../../app/containers/App/Header/PriceDisplay'

import { DEFAULT_CURRENCY_CODE } from '../../app/core/constants'

describe('PriceDisplay', () => {
  test('renders and shows prices', (done) => {
    // Full mount (not shallow) so the snapshot will contain the prices in the rendered html.
    const wrapper = mount(<PriceDisplay neoPrice={25.48} gasPrice={18.10} currencyCode={DEFAULT_CURRENCY_CODE} />)
    expect(wrapper).toMatchSnapshot()
    done()
  })
})
