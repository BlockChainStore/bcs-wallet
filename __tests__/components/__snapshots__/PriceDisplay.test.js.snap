// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceDisplay renders and shows prices 1`] = `
<PriceDisplay
  currencyCode="usd"
  gasPrice={18.1}
  neoPrice={25.48}
>
  <div
    className="navBarItem"
    id="neoPrice"
    key="neoPriceDisplay"
  >
    <span
      className="navBarItemLabel"
    >
      NEO
    </span>
    <span
      className="price"
    >
      $
      25.48
    </span>
  </div>
  <div
    className="navBarItem"
    id="gasPrice"
    key="gasPriceDisplay"
  >
    <span
      className="navBarItemLabel"
    >
      GAS
    </span>
    <span
      className="price"
    >
      $
      18.10
    </span>
  </div>
</PriceDisplay>
`;
